core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -367,7 +367,7 @@ private[deploy] class Master(
             drivers.find(_.id == driverId).foreach { driver =>
               driver.worker = Some(worker)
               driver.state = DriverState.RUNNING
-              worker.drivers(driverId) = driver
+              worker.addDriver(driver)
```

Comment:
From my understanding, `schedule()` will only handle waiting drivers, but here is trying to correctly calculate the exiting drivers, so I don't think `schedule()` will save the issue here. Let me try to test on latest master and back to you the result.